TM-5308, initial fetch and filtering for Cycle Management page

M00staff opened the pull request:
Dual Merge: 
- [API PR](https://github.com/MetaPhase-Consulting/State-TalentMAP-API/pull/886)
- [Mock PR](https://github.com/MetaPhase-Consulting/mock-fsbid/pull/422)

[Ticket](https://metaphase.atlassian.net/browse/TM-5308)
[Ticket](https://metaphase.atlassian.net/browse/TM-4909)

This ticket will only handle the initial data fetch for the ‘Cycle Management’ page, and the filtering of that data (it should be connected to FSBID now)

<img width="1849" alt="image" src="https://github.com/MetaPhase-Consulting/State-TalentMAP-API/assets/12723877/573b9c46-b7db-4656-83d6-433aa9939ccf">

does NOT include adding a new assignment cycle, viewing cycle positions or editing cycle details

M00staff commented on src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasons.jsx:
```diff
@@ -148,12 +143,13 @@ const ManageBidSeasons = () => {
               onChange={setSelectedDates}
-              value={selectedDates}
-              maxDetail="month"
-              calendarIcon={null}
-              showLeadingZeros
+              selected={selectedDates}
+              type="filter"
+              showMonthDropdown
+              showYearDropdown
+              isClearable
```
Updated the Bid Seasons page cause i realized it makes more sense to have a single date (instead of a date range) to filter - this way if either the start date or end date fall within the date selected, it will return those values.

makes WAY more sense this way vs how it was

M00staff replied:
also used a similar pattern for the Cycle Management page, in general - as it's very similar to the Bid Seasons page

M00staff commented on src/Components/CycleManagement/CycleManagement.jsx:
```diff
@@ -2,139 +2,101 @@ import { useEffect, useState } from 'react';
-    selectedCycles,
-    selectedStatus,
-    selectedDates,
+  const filterCyclesByDate = (cycles, date) => {
+    const filteredCycles = cycles.filter(cycle => {
+      const startDate = new Date(cycle.begin_date).getTime();
+      const endDate = new Date(cycle.end_date).getTime();
+      return ((date >= startDate) && (date <= endDate));
+    });
+    return filteredCycles;
```
single date vs date range

M00staff commented on src/Components/CycleManagement/CycleManagement.jsx:
```diff
@@ -2,139 +2,101 @@ import { useEffect, useState } from 'react';
-      setClearFilters(true);
+    let cycles = cycleManagementData;
+    if (selectedCycles.length > 0) {
+      cycles = selectedCycles;
     }
-    if (resetPage) {
-      setPage(1);
+    if (selectedStatus.length > 0) {
+      cycles = cycles.filter(cycle =>
+        selectedStatus.some(status => status.text === cycle.status),
+      );
     }
-    dispatch(saveCycleManagementSelections(getCurrentInputs()));
-    dispatch(cycleManagementFetchData(getQuery()));
-    dispatch(cycleManagementAssignmentCycleFetchData());
+
+    if (selectedDates) return filterCyclesByDate(cycles, selectedDates);
+    return cycles;
```
not sure i LOVE this logic, but it's working pretty well and i think its pretty easily readable

elizabeth-jimenez replied:
i like it 😀

M00staff commented on src/Components/CycleManagement/CycleManagement.jsx:
```diff
@@ -2,139 +2,101 @@ import { useEffect, useState } from 'react';
+    && selectedStatus.flat().length === 0
+    && !selectedDates;
 
-  // Pagination
-  const [page, setPage] = useState(userSelections.page || 1);
-  const [limit, setLimit] = useState(userSelections.limit || 10);
-  const prevPage = usePrevious(page);
-  const pageSizes = POSITION_MANAGER_PAGE_SIZES;
-  const disableInput = addNewCycles || cardsInEditMode.length > 0 || genericFiltersIsLoading;
```
removed all the pagination stuff since we don't have it.

data gets passed into `cycleManagementData$` which is the value we filter & render

M00staff commented on src/Components/CycleManagement/CycleManagement.jsx:
```diff
@@ -2,139 +2,101 @@ import { useEffect, useState } from 'react';
-  const bidCycle = genericFilters$.find(f => f?.item?.description === 'bidCycle');
-  const bidCycleOptions = bidCycle?.data?.length
-    ? nameSort([...new Set(bidCycle.data)], 'name') : [];
+  const uniqueStatuses = () => {
+    const statuses = cycleManagementData.map(cycles => cycles.status);
+    const uniq = [...new Set(statuses)];
+    const uniqObj = uniq.map(x => ({ text: x }));
+    return uniqObj;
+  };
```
so, i could have kept this value hardcoded since there are only 3 statuses, but we usually prefer this way in case there is a value that isn't encapsulated in the dev data that may exist in the prod data. 

could be slicker - but i also think its readable & simple. 

took a minute to realize i needed to pass an object into Picky, and not just an array 😅

elizabeth-jimenez replied:
"but we usually prefer this way in case there is a value that isn't encapsulated in the dev data that may exist in the prod data" 💪

M00staff commented on src/Components/CycleManagement/CycleManagement.jsx:
```diff
@@ -232,67 +195,49 @@ const CycleManagement = (props) => {
+                  value={selectedCycles}
                 />
               </div>
               <div className="filter-div">
-                <div className="label">Cycle:</div>
+                <div className="label">Status:</div>
                 <Picky
                   {...pickyProps}
-                  placeholder="Select Bid Cycle(s)"
-                  options={bidCycleOptions}
-                  valueKey="id"
-                  labelKey="name"
+                  placeholder="Select Status"
+                  options={uniqueStatuses()}
+                  valueKey="text"
+                  labelKey="text"
+                  onChange={setSelectedStatus}
+                  value={selectedStatus}
                   disabled={disableInput}
-                  onChange={setSelectedCycles}
-                  value={selectedCycles}
                 />
               </div>
               <div className="filter-div">
                 <div className="label">Cycle Date:</div>
-                <DateRangePicker
+                <TMDatePicker
                   onChange={setSelectedDates}
-                  value={selectedDates}
-                  maxDetail="month"
-                  calendarIcon={null}
-                  showLeadingZeros
+                  selected={selectedDates}
+                  type="filter"
+                  showMonthDropdown
+                  showYearDropdown
+                  isClearable
```
this is just a re-arrange the filters - makes sense for Cycle to go first since its the most specific

also swapped the date picker for the new one

M00staff commented on src/actions/cycleManagement.js:
```diff
@@ -241,16 +242,6 @@ export function cycleManagementAssignmentCycleFetchData() {
   };
 }
 
-export function cycleManagementSelectionsSaveSuccess(result) {
-  return {
-    type: 'CYCLE_MANAGEMENT_SELECTIONS_SAVE_SUCCESS',
-    result,
-  };
-}
-
-export function saveCycleManagementSelections(queryObject) {
-  return (dispatch) => dispatch(cycleManagementSelectionsSaveSuccess(queryObject));
-}
```
no pagination means we don't need to save the filter selections, as the page load will always load the whole list.

M00staff commented on src/Components/CycleManagement/CycleManagement.jsx:
```diff
@@ -2,139 +2,101 @@ import { useEffect, useState } from 'react';
+    return filteredCycles;
   };
 
-  const getCurrentInputs = () => ({
-    page,
-    limit,
-    selectedCycles,
-    selectedStatus,
-    selectedDates,
-  });
-
-  const getQuery = () => ({
-    'cycle-cycles': selectedCycles.map(bidCycleObject => (bidCycleObject?.id)),
-    'cycle-statuses': selectedStatus.map(statusObject => (statusObject?.code)),
-    'cycle-date-start': isDate(selectedDates?.[0]) ? startOfDay(selectedDates?.[0]).toJSON() : '',
-    'cycle-date-end': isDate(selectedDates?.[1]) ? startOfDay(selectedDates?.[1]).toJSON() : '',
-    limit,
-    page,
-  });
```
no pagination means we dont need any of this for this page

elizabeth-jimenez commented on src/actions/cycleManagement.js:
```diff
@@ -130,6 +130,11 @@ const cyclePosDummyDataToReturn = (query) => new Promise((resolve) => {
+
+let cancelCycleManagementFetch;
+
 export function cycleManagementFetchDataErrored(bool) {
   return {
```
let's clean these files up a bit - nothing fancy, for now, just delete these ones and their corresponding reducers (at a quick glance i dont think we use them):
cyclePositionEditHasErrored
cyclePositionEditIsLoading
cyclePositionEditSuccess

M00staff replied:
i bring this up in the next PR 

right now, I'm just moving all of the actions that get used to the top of the actions file, as i wire them up - so it'll be easier to blow everything unused away, and keep it organized without going through everything individually... it makes sense to me at least  🤷

https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/2813

elizabeth-jimenez replied:
I just don't like the idea of more tickets to handle what should be handled in other tickets. The process when we first did these features was quantity over quality; when Scott and I advocated for cleaner code, the main argument against was 'we can clean it up when we do the integration tickets'. Now that we are at the integration tickets, we are talking about clean-up tickets. My assumption was that as people integrated the features, there would be a holistic approach to also cleaning up the files (bc hopefully thinking about the feature as a whole would make optimizations obvious). I think another PR makes sense, for organization's sake, but not really another ticket.

M00staff replied:
I think I'm following?

My understanding was that we wanted to break this work out into smaller parts so it would be easier to review, move tickets, track progress, etc. - there is definitely cleanup going on, and its happening, when I'm working on the relevant page and it's pointed out in the next PR - but i also made a subtask for going through everything in more detail towards the end.  

There will be a lot of dead code for this feature and I didn't wanna muck up the diff's too much with trying to remove it all at the same time, **especially before I'm sure the code is OK to drop**.  Whether that work is captured in a ticket or subtask is whatever - but a cleanup PR seems neater and easier to review. 

All that being said - its whatever you want - though, it looks to me like `cyclePositionEdit` is used by the `CyclePositionCard` page, and I haven't touched that page yet, or begun work on it.

elizabeth-jimenez replied:
"but a cleanup PR seems neater and easier to review" absolutely 👍

"there is definitely cleanup going on, and its happening, when I'm working on the relevant page" yea, this was all I was trying to get at, and since that's the case, then all good and appreciate it! 

"though, it looks to me like cyclePositionEdit is used by the CyclePositionCard page, and I haven't touched that page yet, or begun work on it" Yea, ur right, it isn't the page ur working on, but i'm always advocating for small changes that improve the project. I'll never ask for things that are 30m+ tho, just small stuff we can quickly sneak in.

So with those Edit redux ones, I think we C/Ped a looot of that initially, but we've established a new pattern now, so I always see removing those as a easy(especially if not being used) win

M00staff replied:
sounds good - yea, just wanted to point out that I'm tracking all the stuff we need to drop, and am very much down for dropping it. 

those actions specifically are used by the page I'm working on now (the edit functionality)... I'll clean them up with that PR

elizabeth-jimenez replied:
perfect 🚀

elizabeth-jimenez commented on src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasons.jsx:
```diff
@@ -31,24 +31,19 @@ const ManageBidSeasons = () => {
     const filteredSeasons = seasons.filter(season => {
       const startDate = new Date(season.bidSeasonsBeginDate).getTime();
       const endDate = new Date(season.bidSeasonsEndDate).getTime();
-      return ((startDate >= startDateRange) && (startDate <= endDateRange))
-        || ((endDate >= startDateRange) && (endDate <= endDateRange));
```
😢

M00staff replied:
well, we never would have gotten to this bit - without that bit, so 🤷

elizabeth-jimenez commented on src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasons.jsx:
```diff
@@ -31,24 +31,19 @@ const ManageBidSeasons = () => {
     if (noFiltersSelected) return ManageBidSeasonsData;
     const seasons = selectedBidSeasons.length > 0 ? selectedBidSeasons : ManageBidSeasonsData;
-    if (selectedDates) return filterSeasonsByDateRange(seasons, selectedDates);
+    if (selectedDates) return filterSeasonsByDate(seasons, selectedDates);
```
update selectedDates to be singular

elizabeth-jimenez commented on src/Components/CycleManagement/CycleManagement.jsx:
```diff
@@ -2,139 +2,101 @@ import { useEffect, useState } from 'react';
   const isSuperUser = userHasPermissions(['superuser'], userProfile?.permission_groups);
 
   // Filters
-  const [selectedCycles, setSelectedCycles] = useState(userSelections?.selectedCycles || []);
-  const [selectedStatus, setSelectedStatus] = useState(userSelections?.selectedStatus || []);
-  const [selectedDates, setSelectedDates] = useState(userSelections?.selectedDates || null);
```
intentional to get rid of the cached selections?

M00staff replied:
yes - its like the bid seasons page, since we don't make a separate call for the filtered data (it just filters the one request) -- there's no need to cache the user selections

elizabeth-jimenez replied:
so the caching of users' selections isnt to save a call, it's to save their inputs - the reason Bid Seasons doesn't use it is bc we don't go anywhere else on that page. The Edit happens on that same page. For Cycle Management the Edit also seems to happen on this page, but then i think View Cycle Positions may take us elsewhere - the hope with the caching is to support the following scenario:

user filters down to five results
user goes away
user comes back to the five results, instead of the whole set

M00staff replied:
updated

elizabeth-jimenez commented on src/Components/CycleManagement/CycleManagement.jsx:
```diff
@@ -2,139 +2,101 @@ import { useEffect, useState } from 'react';
   const [cardsInEditMode, setCardsInEditMode] = useState([]);
   const [addNewCycles, setAddNewCycles] = useState(false);
+  const disableInput = addNewCycles || cardsInEditMode.length > 0;
+  const noFiltersSelected = selectedCycles.flat().length === 0
+    && selectedStatus.flat().length === 0
```
are these ever not flat?

If you wanted, you could do something like this `[selectedCycles, selectedStatus].flat().length === 0`

elizabeth-jimenez replied:
or the spread operator, if they're always flat 🤔

M00staff replied:
either works - went with `[selectedCycles, selectedStatus].flat().length === 0`

elizabeth-jimenez commented on src/Components/CycleManagement/CycleManagement.jsx:
```diff
@@ -2,139 +2,101 @@ import { useEffect, useState } from 'react';
     selectedCycles,
     selectedStatus,
     selectedDates,
+    cycleManagementData,
```
nice catch

elizabeth-jimenez commented:
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/23065001/4d94fd9a-f01f-4152-9454-4aad6af491ac)
maybe dumb question: should status label be moved to the other Closed, Active, Proposed col?

M00staff commented:
> ![image](https://private-user-images.githubusercontent.com/23065001/302726744-4d94fd9a-f01f-4152-9454-4aad6af491ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyNDYzNTksIm5iZiI6MTcwNzI0NjA1OSwicGF0aCI6Ii8yMzA2NTAwMS8zMDI3MjY3NDQtNGQ5NGZkOWEtZjAxZi00MTUyLTk0NTQtNGFhZDZhZjQ5MWFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA2VDE5MDA1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2Nzc3MmE3YWQ2NDc2ZTVlZmNjMGUxYWI5ZjM4MzMzNmIzZDk2NjM3YzczMGRiOTJiOTY2OThkN2M0MDcwYmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0SXYcgtypv_cei09KcResi55UihbLC0fWcCh5-6bArI) maybe dumb question: should status label be moved to the other Closed, Active, Proposed col?

VERY good catch!

Yea - it seems like i have Status and Category conflated ... the mocks actually don't even have the category listed, so i went ahead and removed it, and updated the markup


<img width="1911" alt="image" src="https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/12723877/7a517dc8-e20d-4ef5-bac2-21bb7bb4e7b2">

elizabeth-jimenez commented:
> > ![image](https://private-user-images.githubusercontent.com/23065001/302726744-4d94fd9a-f01f-4152-9454-4aad6af491ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyNDYzNTksIm5iZiI6MTcwNzI0NjA1OSwicGF0aCI6Ii8yMzA2NTAwMS8zMDI3MjY3NDQtNGQ5NGZkOWEtZjAxZi00MTUyLTk0NTQtNGFhZDZhZjQ5MWFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA2VDE5MDA1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2Nzc3MmE3YWQ2NDc2ZTVlZmNjMGUxYWI5ZjM4MzMzNmIzZDk2NjM3YzczMGRiOTJiOTY2OThkN2M0MDcwYmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0SXYcgtypv_cei09KcResi55UihbLC0fWcCh5-6bArI) maybe dumb question: should status label be moved to the other Closed, Active, Proposed col?
> 
> VERY good catch!
> 
> Yea - it seems like i have Status and Category conflated ... the mocks actually don't even have the category listed, so i went ahead and removed it, and updated the markup
> 
> <img alt="image" width="1911" src="https://private-user-images.githubusercontent.com/12723877/302750142-7a517dc8-e20d-4ef5-bac2-21bb7bb4e7b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyNTE4MTUsIm5iZiI6MTcwNzI1MTUxNSwicGF0aCI6Ii8xMjcyMzg3Ny8zMDI3NTAxNDItN2E1MTdkYzgtZTIwZC00ZWY1LWJhYzItMjFiYjdiYjRlN2IyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA2VDIwMzE1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlYTBiYjIwYTJlZmM5ZDk4YzM5MDU0YWQ2MWFlZTQxZTRjZmQ4ZWI0ZTc2YTI4ZjJlNGRkNTNiN2I1NGIwYTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._bS5hEuJzt2JQwLx4-7b_HjC1KEkAKr_mRcM8xRLkQQ">

Nice!

elizabeth-jimenez commented on src/Components/CycleManagement/CycleManagement.jsx:
```diff
@@ -188,7 +157,7 @@ const CycleManagement = (props) => {
 
   // Overlay for error, info, and loading state
-  const noResults = cycleManagementData?.results?.length === 0;
+  const noResults = cycleManagementData$?.length === 0;
```
code will never reach the first `if` of our `getOverlay` - not a call to action just a FWIW

elizabeth-jimenez commented on src/actions/cycleManagement.js:
```diff
@@ -150,6 +155,48 @@ export function cycleManagementFetchDataSuccess(results) {
+  return (dispatch) => {
+    if (cancelCycleManagementFetch) {
+      cancelCycleManagementFetch('cancel');
+      dispatch(cycleManagementFetchDataLoading(true)); // ???
```
not needed cause we never reset for L165 😁

elizabeth-jimenez replied:
wait, i did set it for PP tho 🤔 hmmmmmm down the rabbit hole i go

elizabeth-jimenez replied:
🧠 ur right - it appears we don't need it 🎊

elizabeth-jimenez replied:
im still deep diving this - cancel not quite working how i thought (end goal is the same, but it is taking some deviations)

elizabeth-jimenez replied:
i think i had spun up badly and so i was seeing unexpected behavior - ur right that we dont need to set that loading there! here is the PR with the new pattern https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/2850

you'll want to make sure you include this bit of the pattern to prevent cancels appearing as errors
```
        if (err?.message !== 'cancel') {
          batch(() => {
            dispatch(publishablePositionsSuccess([]));
            dispatch(publishablePositionsErrored(true));
            dispatch(publishablePositionsLoading(false));
          });
        }
```

elizabeth-jimenez commented:
FE Review Complete

Last change requested: `err?.message !== 'cancel'` bit
All other feedback has already been addressed 🎊
